packages/dds/tree/src/test/forest/jsonCursor.ts

```diff
@@ -0,0 +1,148 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Jsonable } from "@fluidframework/datastore-definitions";
+import {
+    ITreeCursor,
+    EmptyKey,
+    FieldKey,
+    TreeNavigationResult,
+    TreeType,
+} from "../../..";
+
+/** NodeTypes used by the JsonCursor. */
+export const enum JsonType {
+    Null = 0,
+    Boolean = 1,
+    Number = 2,
+    String = 3,
+    Array = 4,
+    Object = 5,
+}
+
+/**
+ * An ITreeCursor implementation used to read a JSONable tree for testing and benchmarking.
+ */
+export class JsonCursor<T> implements ITreeCursor {
+    // PERF: JsonCursor maintains a stack of nodes/edges traversed.  This stack is
+    //       partitioned across 3 arrays, with the top of the stack stored in fields.
+    //       This design was advantageous in a similar tree visitor, but should
+    //       be measured again to see if this still provides an advantage.
+
+    private currentNode: any;       // The node currently being visited.
+    private currentKey: FieldKey;   // The key used to navigate from the parent to this node.
+    private currentIndex: number;   // The index used to navigate from the parent to this node.
+
+    private readonly parentStack: any[] = [];       // Ancestors traversed to visit this node.
+    private readonly keyStack: FieldKey[] = [];     // Keys traversed to visit this node, excluding the most recent.
+    private readonly indexStack: number[] = [];     // Indices traversed to visit this node, excluding the most recent.
+
+    constructor(root: Jsonable<T>) {
```

Comment:
I don't think using Jsonable<T> here adds any value over T (since all T are a Jsonable<T>). Seems like making a type for json compatible data (which is not what Jsonable really does...) and using that (and not being generic or using Jsonable) would make more sense.

Reply:
From our chat this morning, Jsonable<T> + type inference is adding some value here, as it can generate compilation errors for things that are provably not JSON compatible.